ask/commands.py

```diff
@@ -0,0 +1,181 @@
+		selected_text = self.view.substr(self.selection)
+		
+		if "." in self.view.file_name():
+			self.langauge = self.view.file_name().split(".")[-1]
+		else: 
+			self.langauge = "txt"
+
+		if not selected_text:
+			return # TODO show error
+
+		if question in description_needed_commands:
+			query = question.format(description=description,code=selected_text)
+		else:
+			query = question.format(code=selected_text)
+		self.view.set_status('Pieces Refactoring', 'Copilot analyzing...')
+		res = pos_client.QGPTApi(PiecesSettings.api_client).question(
+			pos_client.QGPTQuestionInput(
+				query = query,
+				model = PiecesSettings.model_id,
+				relevant = pos_client.RelevantQGPTSeeds(
+					iterable = [
+					#     pos_client.RelevantQGPTSeed(
+					#         seed = pos_client.Seed(
+					#             type="SEEDED_ASSET",
+					#             asset=pos_client.SeededAsset(
+					#                 application=PiecesSettings.application,
+					#                 format=pos_client.SeededFormat(
+					#                     fragment = pos_client.SeededFragment(
+					#                         string = pos_client.TransferableString(raw = selected_text)
+					#                     ),
+					#                 ),
+					#             ), 
+					#         ),
+					#     )
+					]
+				)
+			)
+		)
+		self.view.set_status('Pieces Refactoring', 'Copilot finalizing...')
+		self.window  = self.view.window()
+		response_code = res.answers.iterable[0].text
+		
+		# Regular expression pattern for code block
+		pattern = r'```.*?\n(.*?)```'
+
+		# Find all matches in the markdown text
+		match = re.search(pattern, response_code, re.DOTALL)
+		if match:
+			self.code = match.group(1)
+			self.code_html = self.get_differences(selected_text.splitlines(),self.code.splitlines())
+			link = "<a href=insert>✅ Accept</a> | <a href=insert style='color:red'>❌ Reject</a>"
+			html = f"<div style='display:inline-block'>{link}</div>{self.code_html}"
+
+			# Calculate the length of the code_html
+			code_html_length = len(self.code_html)
+
+			# Create a phantom at the end of the current selection
+			phantom_region = sublime.Region(self.selection.begin(), self.selection.begin() + code_html_length)
+			self.phantom = mdpopups.add_phantom(self.view,"code_phantom", phantom_region, html, sublime.LAYOUT_INLINE,md=False,on_navigate=self.on_nav)
+			self.is_done = True
+			self.view.erase_status('Pieces Refactoring')
+
+
+	def on_nav(self, href):
+		if href == "insert":
+			# Replace the selected text with the code
+			self.view.run_command("replace_selection", {"code": self.code, "selection": [self.selection.a, self.selection.b]})
+			# Remove the phantom
+			mdpopups.erase_phantom_by_id(self.view,self.phantom)
+		elif href == "dismiss":
+			mdpopups.erase_phantom_by_id(self.view,self.phantom)
+
+			
+	def input(self,args):
+		return QuestionInputHandler()
+
+
+	def get_differences(self,s1:list,s2:list):
+
+	    # Compare the snippets
+		diffs = Differ().compare(s1, s2)
```

Comment:
what is this used for? it can potentially be a very expensive operation

Reply:
used to compare 2 snippets 
for example if you asked a refactor it should tell you the difference between the changes that the copilot did before inserting